sdk/src/crypto.rs

```diff
@@ -23,21 +24,70 @@ use crate::{status_code_to_bool, sys, SyscallResult};
 /// NOTE: This only supports f1 and f3 addresses.
 pub fn verify_signature(
     signature: &Signature,
-    signer: &Address,
-    plaintext: &[u8],
+    addr: &Address,
+    digest: &[u8],
```

Comment:
Can we revert the change to this signature as well?  I.e., call `sdk::crypto::hash` inside this function to compute the correct digest.

Reply:
Oh dear. It looks like we're using a special hash function here.

We have two options:

1. One option is to define a multihash codec for this hash function, if it doesn't already exist. Then expose the hash function via the `crypto::hash` syscall.
2. Another option would be to pass the entire messages through instead of just the digests.

I guess the question is: can you think of a reason why a user might want to use a different hash function? Can they use a different hash function for BLS signatures? If not, we should go with option 2 (effectively treating "hashing" as part of signature verification).

We split hashing and verification in Secp256k1 because, in theory, the user may want to use a different hash function. But I _don't_ think that's the case here.

The _downside_ of option two is that the bls signature verification becomes annoying. We'd have to:

1. Concatenate all messages together into a single buffer.
2. Pass a pointer to this buffer along with a pointer to an array of message lengths.

E.g., something like what we do for events:

https://github.com/filecoin-project/ref-fvm/blob/5e743b9235eaaef09097fcd2c7af4e5eea51da96/sdk/src/event.rs#L6-L47

Reply:
For some context: Smart contracts need to be able to verify signatures made by Filecoin wallets, and doing the hashing inside wasm is likely way too expensive.

Reply:
Related: Is it possible to use different curves for BLS signatures? Does anyone actually _use_ a different curve?

Reply:
> Can they use a different hash function for BLS signatures?

A BLS signer can use a different hash-to-curve algorithm than the default one used in `bls-signatures` (i.e. a signer may use any hash function that uniquely maps a string of bytes to a G2 point of curve BLS12-381), however hash-to-curve algorithms are generally considered a hard thing to do correctly and (in my experience) a single hash-to-curve algorithm is chosen and hardcoded into a protocol.

It looks like the IETF hash-to-curve standard suggests [just the one hash function for curve BLS12-381](https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-hash-to-curve-12#name-bls12-381-g2).

> can you think of a reason why a user might want to use a different hash function? 

The only thing I can think of is a user wanting to use a gas cost optimized hash function, however I don't know if a secure hash function with that property exists.

Reply:
> 1. Concatenate all messages together into a single buffer.
> 2. Pass a pointer to this buffer along with a pointer to an array of message lengths.

From the PR comments above, I was under the impression that SDK's `verify_signature` should operate on single-message signatures (i.e. Secp and non-aggregated BLS) and `verify_bls_aggregate` used for multi-message signatures. I think that you are referring to just `verify_bls_aggregate` here?

Reply:
Ok, if it's standard, I'd say that we make "hashing" the responsibility of the syscall.

> From the PR comments above, I was under the impression that SDK's verify_signature should operate on single-message signatures (i.e. Secp and non-aggregated BLS) and verify_bls_aggregate used for multi-message signatures. I think that you are referring to just verify_bls_aggregate here?

Correct. Basically:

1. `verify_signature` must take the message, not the digest.
2. Given that `verify_signature` takes the message, how can we implement `verify_bls_aggregate`. I.e., `verify_bls_aggregate` needs to either (a) hash inside the SDK (actor) or (b) hash inside the FVM.

Reply:
> 1. Concatenate all messages together into a single buffer.
> 2. Pass a pointer to this buffer along with a pointer to an array of message lengths.

Rather that copying the plaintexts at the SDK level (i.e. concatenating), couldn't the SDK create an array of plaintext slices `[&[u8]]` (a bytes slice is 16 bytes, which is probably smaller than each plaintext), then pass to the syscall a pointer `*const &[u8] as *const u8` to the allocated array of plaintext slices:

```rust
// SDK
pub fn verify_signature(
    signature: &Signature,
    addr: &Address,
    plaintext: &[u8],
) -> SyscallResult<bool> {
    match sig_type {
        SignatureType::BLS => {
            let sig = sig_bytes.try_into();
            let pub_key = addr.payload();
            // Pass slice of paintext slices to syscall.
            verify_bls_aggregate(sig, &[pub_key], &[plaintext])
        }
        SignatureType::Secp256k1 => {
             let sig = sig_bytes.try_into();
             let digest = hash_blake2b(plaintext);
             let addr_recovered =
                 recover_secp_public_key(digest, sig).map(Address::new_secp256k1)?;
            Ok(addr == &addr_recovered)
        }
}

// SDK
pub fn verify_bls_aggregate(
    sig: &[u8; BLS_SIG_LEN],
    pub_keys: &[[u8; BLS_PUB_LEN]],
    plaintexts: &[&[u8]],
) -> SyscallResult<bool> {
    unsafe {
        sys::crypto::verify_bls_aggregate(
            pub_keys.len() as u32,
            sig.as_ptr() as *const u8,
            pub_keys.as_ptr() as *const u8,
            // Cast `*const &[u8]` to `*const u8`.
            plaintexts.as_ptr() as *const u8,
        )
        .map(status_code_to_bool)
    }
}

// Syscall
pub fn verify_bls_aggregate(
    context: Context<'_, impl Kernel>,
    num_signers: u32,
    sig_off: u32,
    pub_keys_off: u32,
    plaintext_slices_off: u32,
) -> Result<i32> {
    let sig: &[u8; BLS_SIG_LEN] = context.try_slice(...);
    let pub_keys: &[[u8; BLS_PUB_LEN]] = context.try_chunks(...);

    // 16 bytes per slice
    let plaintext_slices_len = num_signers * std::mem::size_of::<&[u8]>();
    let plaintext_slices: &[&[u8]] = context
        .memory
        .try_slice(plaintext_slices_off, plaintext_slices_len)
        .map(|plaintext_slices| unsafe {
            std::slice::from_raw_parts(
                plaintext_slices.as_ptr() as *const &[u8],
                num_signers,
            )
        })?;

     context
        .kernel
        .verify_bls_aggregate(&sig, &pub_keys, plaintexts)
        .map(|v| if v { 0 } else { -1 })
}
```

then the kernel performs plaintext hashing in `CryptoOps::verify_bls_aggregate` via `shared::crypto::signature::verify_bls_aggregate`?

Reply:
I did this technique in an earlier commit to pass a slice of digest array references `&[&[u8; BLS_DIGEST_LEN]]` from the SDK to syscall without copying the digests (at either the SDK or FVM levels).

Reply:
Hm. So, the difference is that the "slices" here are fat pointers (that's why they're 16 bytes). And the lengths are dynamic. And, unfortunately, the layout is undefined IIRC so we can't just pass it through:

So, we _could_ likely map the input slice as:

```rust
plaintexts.map(|ptr| [ptr.as_ptr(), ptr.len()]).collect()
```

But this may be more annoying to charge for in terms of gas (and will require an allocation). That is, in the FVM, we'd have to:

1. In the _syscall binding_, we'd have to slice N times (and store those slices in a new vec that we'd have to allocate FVM-side). We can get away with charging nothing for a small constant amount of work (we can consider this to be "covered" by the base syscall gas fee) but, if we perform a variable amount of work, we need to charge for that.
2. Then, in the _kernel_, we'd have to charge for the actual hashing, etc.

If we instead pass a concatenated buffer through, this becomes a bit simpler:

1. We have to slice once in the syscall binding (and don't have to copy anything or charge any gas).
2. We can charge gas all at once in the kernel for hashing, etc. before we do any actual work.

The downside is that we have more copying in the actor, but there are bulk copy instructions in wasm which should compile down to some pretty-well optimized assembly, so it should be fast. We charge 0.4gas/byte copied, which is likely less than a percent of the charge of hashing.

Reply:
@DrPeterVanNostrand have you had a chance to take a look at this?

Reply:
Sorry, I had proofs' work last week then was afk for a couple of days. I picked it back up today and am currently adding the plaintext concatenation.

Reply:
Thanks! And if this slips, it's not the end of the world compared to everything else coming in nv21. This can ship in nv22 without anyone complaining.